src/providers/ibmcloud/mod.rs

```diff
@@ -0,0 +1,82 @@
+//! Metadata fetcher for the IBMCloud provider.
+//!
+//! IBMCloud supports multiple kind of compute nodes, with different
+//! features and peculiarities.
+
+// TODO(lucab): this allows adding 'classic' and 'vpc-gen1' instances too
+//  via auto-detection, if there is a need for that in the future.
+
+use crate::errors::Result;
+use crate::{network, providers};
+use openssh_keys::PublicKey;
+use std::collections::HashMap;
+
+mod gen2;
+
+/// IBMCloud provider (all supported flavors).
+#[derive(Debug)]
+pub enum IbmCloudProvider {
+    /// VPC Generation 2.
+    Gen2(gen2::G2Provider),
+}
+
+impl IbmCloudProvider {
+    /// Build a new client for IBMCloud.
+    ///
+    /// This internally tries to autodetect the infrastructure type
+    /// and mount the relevant config-drive.
+    pub fn try_autodetect() -> Result<Self> {
+        // Auto-detection order: Gen2 (possibly later: classic, Gen1).
+        slog_scope::trace!("trying to autodetect ibmcloud infrastructure type");
+
+        match gen2::G2Provider::try_new() {
```

Comment:
Have a question here:
Is it that if we failed to create a `gen2::G2Provider` (failed to create a temp dir or mount the file system), then the program failed to auto-detect the infrastructure type? I'm having this question because the 'detection' step is not obvious to me.
Otherwise LGTM! :+1:

Reply:
The detection logic is meant to short-circuit (i.e. return an `Ok(...)`) in the positive case, and otherwise log an error and continue by trying other sub-types in order.
That is, there will be other `match` statements after this one for the other sub-types.

Overall detection only fails when all types have failed, and we reach the `bail!()` statement at the bottom.

Self-note: we'll need to have some scatter-gather parallelism here, otherwise auto-detection will take a linear amount of time with the number of sub-types.